src/main/scala/com.snowplowanalytics.sauna/responders/hipchat/SendRoomNotificationResponder.scala

```diff
@@ -0,0 +1,238 @@
+
+  implicit val selfDescribingReads: Reads[SelfDescribing] = (
+    (JsPath \ "schema").read[String].map[SchemaKey](schemaKey => SchemaKey.fromUri(schemaKey).get) and
+      (JsPath \ "data").read[JsValue]
+    ) (SelfDescribing.apply _)
+
+  /**
+   * A Sauna command.
+   *
+   * @param envelope Identifies the command and provides instructions on how to execute it.
+   * @param command  Specifies the exact action to execute.
+   */
+  case class SaunaCommand(
+    envelope: SelfDescribing,
+    command: SelfDescribing)
+
+  implicit val saunaCommandReads: Reads[SaunaCommand] = (
+    (JsPath \ "envelope").read[SelfDescribing] and
+      (JsPath \ "command").read[SelfDescribing]
+    ) (SaunaCommand.apply _)
+
+  /**
+   * TODO: define what this is
+   */
+  object Semantics extends Enumeration {
+    type Semantics = Value
+    val AT_LEAST_ONCE = Value("AT_LEAST_ONCE")
+  }
+
+  implicit val semanticsReads: Reads[Semantics] = Reads.enumNameReads(Semantics)
+
+  /**
+   * TODO: define what this is
+   *
+   * @param semantics  TODO: define what this is
+   * @param timeToLive The command's lifetime (in ms): if a command is
+   *                   received after `timeToLive` ms from its' creation have
+   *                   passed, it will be discarded. (If None, a command won't be
+   *                   discarded based on creation time.)
+   */
+  case class ExecutionParams(
+    semantics: Semantics,
+    timeToLive: Option[Int]
+  )
+
+  implicit val timeToLiveReads: Reads[Option[Int]] = JsPath.readNullable[Int]
+  implicit val executionParamsReads: Reads[ExecutionParams] = Json.reads[ExecutionParams]
+
+  /**
+   * A Sauna command envelope.
+   *
+   * @param commandId   A unique identifier (uuid4) for the command.
+   * @param whenCreated The command's creation date.
+   * @param execution   TODO: define what this is
+   * @param tags        TODO: define what this is
+   */
+  case class CommandEnvelope(
+    commandId: String,
+    whenCreated: LocalDateTime,
+    execution: ExecutionParams,
+    tags: Map[String, String]
+  )
+  implicit val commandEnvelopeReads: Reads[CommandEnvelope] = Json.reads[CommandEnvelope]
+
+  /**
+   * Constructs a [[Props]] for a [[SendRoomNotificationResponder]] actor.
+   *
+   * @param hipchat The HipChat API wrapper.
+   * @param logger  A logger actor.
+   * @return [[Props]] for the new actor.
+   */
+  def props(hipchat: Hipchat, logger: ActorRef): Props =
+    Props(new SendRoomNotificationResponder(hipchat, logger))
+
+  /**
+   * Attempts to extract a Sauna command from a [[JsValue]].
+   *
+   * @param json The [[JsValue]] to extract a command from.
+   * @tparam T The type of the command's data.
+   * @return Right containing a tuple of the command's envelope and data
+   *         if the extraction was successful, Left containing an error message
+   *         otherwise.
+   */
+  def extractCommand[T](json: JsValue)(implicit tReads: Reads[T]): Either[String, (CommandEnvelope, T)] = {
```

Comment:
Whole pattern-matching thing can be rewritten as something like:

```scala
    for {
      s <- json.validate[SelfDescribing]
      c <- s.data.validate[SaunaCommand]
      e <- c.envelope.data.validate[CommandEnvelope]
      d <- c.command.data.validate[T]
    } yield (e, d)   // This is JsResult[(CommandEnvelope, T)]
```

This beautiful magic is possible because `JsResult` is a monad and supports `for`-comprehensions. Error-message in Left should be derived from `Reads` instance of particular decoder. 

If you need further explanations on why this pattern-matching can be replaced with `for`-comprehension - you can ask me in HipChat.